---
title: 'Running the gene-regulation pipelines on a virtual machine on the IFB cloud'
author: "Claire Rioualen"
date: "May 12th, 2016"
output:
  html_document:
    fig_caption: yes
    highlight: zenburn
    number_sections: yes
    theme: cerulean
    toc: yes
    toc_depth: 4
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    number_sections: yes
    toc: yes
    toc_depth: 4
  word_document:
    toc: yes
    toc_depth: 4
---

<!--
- [1. Using the Gene-regulation appliance](#1-using-the-gene-regulation-appliance)
    - [1.1 Requirements](#11-requirements)
    - [1.2 Virtual disk creation](#12-virtual-disk-creation)
    - [1.3 Creation of an instance](#13-creation-of-an-instance)
    - [1.4 Connection to the device](#14-connection-to-the-device)
    - [1.5 Download source data](#15-download-source-data)
    - [1.6 Execute workflow](#16-execute-workflow)
- [2. Visualizing results](#2-visualizing-results)
    - [2.1 Install and run the X2Go client on your host computer](#21-install-and-run-the-x2go-client-on-your-host-computer)
    - [2.2 Visualize results](#22-visualize-results)
        - [FastQC](#fastqc)
        - [IGV](#igv)
- [3. Create your own Gene-regulation appliance](#3-create-your-own-gene-regulation-appliance)
    - [Creation of an instance](#creation-of-an-instance)
    - [Installing programs and dependencies](#installing-programs-and-dependencies)
    - [Get the `gene-regulation` repository](#get-the-gene-regulation-repository)
    - [Run makefile to install the dependencies](#run-makefile-to-install-the-dependencies)
-->

> Notes: 
> This protocol was developed on a Unix computer, with the OS LMDE. 
> The virtual machines are developed with the Ubuntu 14.04 OS. 
> The protocol is based on gene-regulation v1.0

# 1. Using the Gene-regulation appliance

## 1.1 Requirements

**User account creation & configuration**

* Using the IFB cloud facilities requires to have a user account. Register [here](https://cloud.france-bioinformatique.fr/accounts/register/). 

* Once your account has been validated, you can [login](https://cloud.france-bioinformatique.fr/accounts/login/).

* In order to be able to access your instances through SSH, you should register your SSH public key in your [account settings](https://cloud.france-bioinformatique.fr/cloud/profile/), through the dashboard.

## 1.2 Virtual disk creation

Appliances usually have a limited amount of disk space (up to 10, 20Go). If the instance to be run necessitates disk space, you have to create a virtual disk (vDisk) prior to launching it. By default, the capacity of storage granted to a user is 250Go, which can be divided into as many vDisks as necessary. When instantiating an appliance, you can chose to attach one of these vDIsks to the virtual machine. You'll be able to access data on this disk through SSH. 

1. Click *New vDisk* button.
2. Enter a size (whole number equating to the amount of Go needed). 
3. Name it (e.g. `GSE20870-10Gb`, the ID of the Gene Expression Omnibus series that will be stored on the virtual drive).

![](../../img/vdisk-x2go.png)

<!--\includegraphics[width=250pt]{img/vdisk-x2go.png}-->


## 1.3 Creation of an instance

1. Click *New Instance* button.
2. Choose appliance "Ubuntu 14.04 IFB-X2GO-10GB" in the drop-down menu. TODO Change
3. Name your VM.
4. Choose the amount of CPU and RAM to grant the VM (up to 8 CPU, 32 GB RAM).
5. Attach the vDisk.
6. Click *Run*.

![](../../img/ubuntu_x2go_create.png)

7. After a few seconds, you may refresh the page until the newly created instance shows up on the dashboard. Clicking on the ssh mention in the *Access* column will give you the commands to access your virtual machine. 

![](../../img/x2go_ssh.png)

## 1.4 Connection to the device

Open a terminal on your host computer and type in:
```
ssh -A -p 22 root@192.54.201.124
```

## 1.5 Download source data

On the IFB cloud VM, the vDisk is automatically attached and mounted by default under `/root/mydisk`, with a link from your home folder `~/mydisk`. 

Here we create a folder to store the source data files and download them .

```
## Download input reads
mkdir -p ~/mydisk/GSE20870_analysis/source/GSE20870/GSM521934 
wget -v --no-clobber ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX%2FSRX021%2FSRX021358/SRR051929/SRR051929.sra -P ~/mydisk/GSE20870_analysis/source/GSE20870/GSM521934/

## Download ChIP-seq reads
mkdir -p ~/mydisk/GSE20870_analysis/source/GSE20870/GSM521935
wget -v --no-clobber ftp://ftp-trace.ncbi.nlm.nih.gov/sra/sra-instant/reads/ByExp/sra/SRX%2FSRX021%2FSRX021359/SRR051930/SRR051930.sra -P ~/mydisk/GSE20870_analysis/source/GSE20870/GSM521935/
```


<!--
For convenience, we create a symlink from our preferred location `/data` to the data directory `~/mydisk/GSE20870_analysis`. 

```
ln -s ~/mydisk/GSE20870_analysis ~/data
```
-->

You can use your preferred locations by updating the workflow configuration file located in `gene-regulation/examples/GSE20870/GSE20870.yml`. 

## 1.6 Execute workflow

```
cd  ~/mydisk/GSE20870_analysis/
ln -s ~/gene-regulation-1.0 ./gene-regulation
snakemake -p -s gene-regulation/scripts/snakefiles/workflows/factor_workflow.py --configfile gene-regulation/examples/GSE20870/GSE20870.yml
```

Using 4CPU & 8Go of RAM, the workflow took about 12mn to complete. 

Congratulations! You just executed this wonderful workflow:

![](../../img/rule.png)



# 2. Visualizing results

## 2.1 Install and run the X2Go client on your host computer


The Virtual Machine created on the IFB cloud doesn't have a graphical interface, but it contains the X2GO software. We're gonna use it to create a distant desktop to visualize the results from the host machine. 

1. Install the x2go client and launch it from your local computer.

```
sudo apt-get install x2goclient
x2goclient
```

<!--2. Copy your ssh key to the authorized keys of the virtual machine. (**Ã  revoir !!**)

```
cat $HOME/.ssh/id_rsa.pub | ssh root@192.54.201.124 "cat >> .ssh/authorized_keys"
```
-->

2. Create a new session using the Mate desktop.

![](../../img/x2goclient_session_create.png)

3. The session now appears on the right panel. Just click it to lauch it!

![](../../img/x2go_launch_session.png)

4. You should be now on the virtual desktop! 

![](../../img/mate_term.png)

Note: you may need to change your keyboard settings 

* Go to **System** > **Preferences** > **Keybords**
* Click on tab **Layouts**
* Add and/or remove desired keyboards

## 2.2 Visualize results

### FastQC

You can visualize the FastQC results using firefox or any other navigator. Fetch the `html` files located in the sample directories.

* Before trimming:
```
firefox /data/results/GSE20870/samples/GSM521934/GSM521934_fastqc/GSM521934_fastqc.html
firefox /data/results/GSE20870/samples/GSM521935/GSM521935_fastqc/GSM521935_fastqc.html
```
* After trimming:
```
firefox /data/results/GSE20870/samples/GSM521934/GSM521934_sickle-se-q20_fastqc/GSM521934_sickle-se-q20_fastqc.html
firefox /data/results/GSE20870/samples/GSM521935/GSM521935_sickle-se-q20_fastqc/GSM521935_sickle-se-q20_fastqc.html
```


![](../../img/x2go_fastqc.png)

### IGV

You can visualize the peaks by running IGV from the terminal. You may need to source the `~/.bashrc` first in order to update teh `$PATH`. 
```
source ~/.bashrc
igv
```

* Click "File" > "Open session..." and chose the file `/data/results/GSE20870/peaks/igv_session.xml`.
* You may need to adjust the panel sizes. 


![](../../img/igv.png)


# 3. Create your own Gene-regulation appliance

Creating a new appliance from scratch is very similar to using one. You have to satisfy the requirements described in part 1.1. 

If you want to manipulate data, you should also create a vDisk following step 1.2.

## Creation of an instance

When creating a new instance choose a 10 Go Ubuntu appliance and check the "Create appliance" option:

1. Click *New Instance* button.
**2. Choose appliance "Ubuntu 14.04 IFB-X2GO-10GB" in the drop-down menu. **
3. Name your VM.
4. Choose the amount of CPU and RAM to grant the VM (up to 8 CPU, 32 GB RAM).
**5. Check the box *Create appliance*.**
6. Attach the vDisk.
7. Click *Run*.

![](../../img/create_appliance.png)

The new instance should appear in orange bold fonts in the dashboard. 

![](../../img/ubuntu_create.png)
 
You can connect to the instance through `ssh` as shown in part 1.4.

# Installing programs and dependencies

Once in the virtual machine, you can install the required programs. 

## Get the `gene-regulation` repository

```
wget https://github.com/rioualen/gene-regulation/archive/1.0.tar.gz
tar zvxf 1.0.tar.gz
cd gene-regulation-1.0
```



## Run makefile to install the dependencies 

This may take a while (up to 30mn-1h) & source the `.bashrc` in order to update the `$PATH` accordingly.

```
make -f scripts/makefiles/install_tools_and_libs.mk all
source ~/.bashrc
```

You should now be able to execute the example workflow by following steps 1.5 and 1.6.

In order for your appliance to remain persistant and be available to other users on the IFB cloud, you should contact an admin: @?
