# File formats

*NB* Should be merged with glossary section?

A list of formats maintained by the [UCSC](http://genome.ucsc.edu/FAQ/FAQformat.html).

 * **fastq:** raw sequences + quality ([more](http://maq.sourceforge.net/fastq.shtml)).
 * **sam:** aligned reads ([more](http://bioinformatics.oxfordjournals.org/content/25/16/2078.full.pdf)).
 * **bam:** compressed sam ([more](http://bioinformatics.oxfordjournals.org/content/25/16/2078.full.pdf)).
 * **cigar:** alignment ([more]()).
 * **bed:** genomic coordinates ([more](http://genome.ucsc.edu/FAQ/FAQformat.html#format1)).
 * **wig:** coverage / density of some signal along genome ([more](http://genome.ucsc.edu/goldenPath/help/wiggle.html)).
 * **bedgraph:** positions + density values ([more](http://genome.ucsc.edu/goldenPath/help/bedgraph.html)).
 * **gff:** genome feature file - annotations ([more](http://www.sanger.ac.uk/resources/software/gff/spec.html)).
 * **gtf:** variant of GFF, with two fields for annotation ([more](http://www.ensembl.org/info/website/upload/gff.html)).
 * **gft2:** Gene annotation ([more](http://mblab.wustl.edu/GTF22.html)).
 * **fastq:** raw reads + quality scores ([more](http://maq.sourceforge.net/fastq.shtml)).
 * **pileup:** base-pair information at each chromosomal position ([more](http://samtools.sourceforge.net/pileup.shtml)).
 * **vcf:** variant call format ([more](http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41)).