# Resources

## Research articles

## Protocols

## ChIP-seq guidelines

* [Bailey et al., 2013](http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003326). Practical Guidelines for the Comprehensive Analysis of ChIP-seq Data.

* [ENCODE & modENCODE consortia, 2012](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3431496/). ChIP-seq guidelines and practices of the ENCODE and modENCODE consortia.

## Tutorials

### French

* [Thomas-Chollier et al. 2012](http://www.nature.com/nprot/journal/v7/n8/full/nprot.2012.088.html). A complete workflow for the analysis of full-size ChIP-seq (and similar) data sets using peak-motifs.
* **TODO** add JvH & MTC tutos
* **TODO** Roscoff bioinformatics school: [link](http://ecole-bioinfo-aviesan.sb-roscoff.fr/archives-2014)
* [RNA-seq tutorial](http://bioinfo-fr.net/lanalyse-de-donnees-rna-seq-mode-demploi)

### English

* [Galaxy tutorial](https://wiki.galaxyproject.org/Learn)

## Databases

* [Wikipedia list](https://en.wikipedia.org/wiki/List_of_biological_databases)


