# NGS analysis tools

This list is far from exhaustive. You can find other lists:

* [Sequencing (OmicTools)](http://omictools.com/sequencing-c152-p1.html)
* [ChIP-seq (OmicTools)](http://omictools.com/chip-seq-c1215-p1.html)
* [ChIP-seq](https://github.com/crazyhottommy/ChIP-seq-analysis)

## Quality assessment

### FastQC

[Link](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/)

FastQC aims to provide a simple way to do some quality control checks on raw sequence data coming from high throughput sequencing pipelines. It provides a modular set of analyses which you can use to give a quick impression of whether your data has any problems of which you should be aware before doing any further analysis.

The main functions of FastQC are:

* Import of data from BAM, SAM or FastQ files (any variant)
* Providing a quick overview to tell you in which areas there may be problems
* Summary graphs and tables to quickly assess your data
* Export of results to an HTML based permanent report
* Offline operation to allow automated generation of reports without running the interactive application


## Trimming

The quality of the reads generated by high-throughput sequencing technologies tend to decrease at their ends. Trimming consists in cutting out theses ends, and thus better the quality of reads before the mapping.

### Sickle

[Link](https://github.com/najoshi/sickle)

Sickle uses sliding windows computing sequencing quality along the reads. When the quality falls below a chose q-value threshold, the reads is cut. If the size of the remaining read is too short, it is completely removed. 
Sickle takes into account three different types of read quality: Illumina, Solexa, Sanger. 

### Cutadapt

## Alignment/mapping

*note* find that presentation explaining the difference between the 2 of em

*to translate*

Le but de l'alignement est de replacer les //reads// issus du séquençage à leur emplacement sur un génome de référence. Lorsque le //read// est suffisamment long, il peut généralement être //mappé// sur le génome avec une bonne certitude, en tolérant une certain quantité de //mismatches//, c'est-à-dire de nucléotides mal appariés. Néanmoins certaines séquences répétées du génome peuvent s'avérer plus difficiles à aligner. 
On désigne par l'expression "profondeur de séquençage" (ou //sequencing depth//) le nombre moyen de //reads// alignés par position sur le génome. Plus cette profondeur est importante, meilleure est la qualité de l'alignement, et plus les analyses ultérieures seront de qualité. 

### BWA

[Link](http://bio-bwa.sourceforge.net/)

[Manual](http://bio-bwa.sourceforge.net/bwa.shtml)

[Publication](http://www.ncbi.nlm.nih.gov/pubmed/19451168): Li H. and Durbin R. (2009). Fast and accurate short read alignment with Burrows-Wheeler Transform. Bioinformatics, 25:1754-60.

### Bowtie

### Others 

A list on [Wikipedia](https://en.wikipedia.org/wiki/List_of_sequence_alignment_software)

## Peak-calling

**TODO** section à étoffer, voir protocole d'install snakemake.

### MACS14/MACS2
### SWEMBL
### HOMER (findPeaks)
### SPP
### bPeaks
### SICER

## Motif analysis

### Regulatory Sequence Analysis Tools (RSAT)

[Link](http://rsat.eu/)

*to translate*

Suite logicielle spécialisée pour l'analyse de motifs cis-régulateurs, développée par les équipes de Morgane Thomas-Chollier (ENS, Paris) et Jacques van Helden (TAGC, Marseille). Inclut des outils spécifiques pour l'analyse de données de ChIP-seq.

More: see the tutorials section in `resources.md`.

### MEME

[Link](http://meme.ebi.edu.au/meme/doc/meme-chip.html)

*to translate*

Suite logicielle spécialisée pour l'analyse de motifs cis-régulateurs, développée par l'équipe de Tim Bailey. Inclut des outils spécifiques pour l'analyse de données de ChIP-seq. 



## File conversion

### SamTools

[Link](http://samtools.sourceforge.net/)

### BamTools

### BedTools

[Link](http://bedtools.readthedocs.org/en/latest/)

### SRA Toolkit

[Documentation](http://www.ncbi.nlm.nih.gov/Traces/sra/?view=toolkit_doc)

Set of tools for the conversion of `*.sra` files (sequence read archive) into several formats. `fastq-dump` converts to `*.fastq` files.

* More info on the [SRA format](http://www.ncbi.nlm.nih.gov/Traces/sra/)
* fastq-dump [manual](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump)
* [Download](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=software)
* [Install guide](http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=std)

## Miscellaneous

* [MICSA](http://bioinfo-out.curie.fr/software.html): peak-calling & motifs discovery ([publication](http://bioinformatics.oxfordjournals.org/content/26/20/2622.long)).
* [ChIPMunk](http://line.imb.ac.ru/ChIPMunk): deep and wide digging for binding motifs in ChIP-Seq data ([publication](http://bioinformatics.oxfordjournals.org/content/26/20/2622.long)).
* [HMCan](http://www.cbrc.kaust.edu.sa/hmcan/): a method for detecting chromatin modifications in cancer samples using ChIP-seq data ([publication](http://bioinformatics.oxfordjournals.org/content/29/23/2979.long)).
* seqMINER
* [Crunch project](http://crunch.unibas.ch/fcgi/crunch.fcgi)
* CSDeconv
* ...

