"""
Download genome sequence to location specified in the config file.

Assumes you provided a url corresponding to a .fa.gz file.

Usage example:

Required parameters:
    config["qsub"]
    config["dir"]["genome"]
    config["genome"]["version"]
    config["genome"]["gff3_url"]

Optional parameters:

Authors: 
    Claire Rioualen
"""

# Set parameters
if not "qsub" in config.keys():
    sys.exit("The parameter qsub must be defined in the config file")

if not "version" in config["genome"].keys():
    sys.exit("Genome version must be defined in the config file")

if not "genome" in config["dir"].keys():
    sys.exit("Genome dir must be defined in the dir section of the config file")

if not "gff3_url" in config["genome"].keys():
    sys.exit("The parameter genome gff3_url must be defined in the config file")

# Define paths
genome_version = config["genome"]["version"]
genome_dir = config["dir"]["genome"] + genome_version + "/"
genome_file = genome_dir + genome_version + ".gff3"

rule annotation_download:
#    input: "."
    output: genome_file
    log: genome_file + ".log"
    benchmark: genome_file + "_benchmark.json"
    params:
        gff = config["genome"]["gff3_url"], \
        dir = genome_dir, \
        version = genome_version, \
        qsub = config["qsub"] \
          + " -e " + genome_dir + "download_gff_qsub.err" \
          + " -o " + genome_dir + "download_gff_qsub.out"
    shell:"""
(cd {params.dir}
wget {params.gff}
filename=$(basename {params.gff})
echo $filename
gunzip $filename
fastaname=$(basename $filename .${{filename##*.}})
mv $fastaname {params.version}.gff3) &> {log}
"""

