rule bed_to_fasta:
	"""Fetch sequences from UCSC to obtain a fasta file from a set of
	genomic coordinates described in a bed file.

	Example: 
	mkdir -p test/fetch_seq; cd test/fetch_seq; 
	wget http://pedagogix-tagc.univ-mrs.fr/rsat/demo_files/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.bed; 
	cd -
	snakemake --snakefile ${RSAT}/snakemake_files/chip-seq_motifs.py test/fetch_seq/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.fasta

	Required parameters:
		config["qsub"]
		config['genome']['genome_version']

	Usage example: 
		input:expand('{read_dir}/{dataset}_{aligner}.fasta', read_dir=READS, sample=DATASET, aligner=ALIGNERS)
	"""
	input: "{peaks}.bed"
	output: "{peaks}.fasta"
	log: "{peaks}_bed_to_fasta.log"	
	benchmark: "{peaks}_bed_to_fasta_benchmark.json"
	params: 
		genome = config['genome']['genome_version'], \
		qsub = config["qsub"] + " -q short -e {peaks}_bed_to_fasta_qsub.err -o {peaks}_bed_to_fasta_qsub.out"
	shell:"fetch-sequences -i {input} -header_format galaxy -genome {params.genome} -o {output} 2> {log}"
