

rule bed_to_fasta:
	"""Fetch sequences from UCSC to obtain a fasta file from a set of
	genomic coordinates described in a bed file.

	Example: 
	mkdir -p test/fetch_seq; cd test/fetch_seq; 
	wget http://pedagogix-tagc.univ-mrs.fr/rsat/demo_files/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.bed; 
	cd -
	snakemake --snakefile ${RSAT}/snakemake_files/chip-seq_motifs.py test/fetch_seq/fetch-sequences_Schmidt_2011_mm9_CEBPA_SWEMBL_R0.12_702peaks.fasta
	"""
	input: RESULTSDIR + "{CHIP}_vs_{INPUT}/{CHIP}_{ALIGNER}_{PEAK_CALLERS}.bed"
	output: RESULTSDIR + "{CHIP}_vs_{INPUT}/{CHIP}_{ALIGNER}_{PEAK_CALLERS}.fasta"
	log: RESULTSDIR + "{CHIP}_vs_{INPUT}/{CHIP}_{ALIGNER}_{PEAK_CALLERS}_fetch-seq.log"								## Empty files?
	benchmark: RESULTSDIR + "{CHIP}_vs_{INPUT}/{CHIP}_{ALIGNER}_{PEAK_CALLERS}_fetch-seq_benchmark.json"
#	params: qsub = QSUB_PARAM + " -q short -e {reads}_bam_to_bed_qsub.err -o {reads}_bam_to_bed_qsub.out"
	shell:"fetch-sequences -i {input} -header_format galaxy -genome {GENOME} -o {output} 2> {log} "
