rule bowtie:
	"""Read mapping with bowtie version 1 (no gap).

	   Bowtie 1 may be faster than bowtie2 in some configurations, for
	   example when reads are mapped onto small genomes without gap
	   (e.g. Bacteria).
	   
	   Warning: to use this rule you first need to trim your reads by
	   quality score. We use fastqc to check read quality, and sickle for
	   trimming.

	   This rule requires to specify the following parameters: 
		config["bowtie"]["max_mismatches"]
		config["bowtie"]["reads_iso"]
		config["bowtie"]["bowtie_index"]
		config["qsub"]
	    These parameters have to be defined in the configuration file (in
	    json format).

	"""
	input: reads = "{reads}.fastq", index1 = config["bowtie"]["bowtie_index"] + ".1.ebwt"
	output: sam = "{reads}_bowtie.sam", summary = "{reads}_bowtie_summary.txt"
	log: "{reads}_bowtie.log"
	benchmark: "{reads}_bowtie_benchmark.json"
	params: max_mismatches = config["bowtie"]["max_mismatches"], \
		reads_iso = config["bowtie"]["reads_iso"], \
		index = config["bowtie"]["bowtie_index"], \
		qsub = config["qsub"] + " -q long -e {reads}_bowtie_qsub.err -o {reads}_bowtie_qsub.out"
	shell: "bowtie -t -q -S -n {params.max_mismatches} -m {params.reads_iso} --phred33-quals {params.index} {input.reads} {output.sam} > {output.summary} 2> {log}"


# """
#    Read mapping with bowtie version 1 (no gap, no paired-end).
   
#    Warning: to use this rule you first need to trim your reads by
#    quality score. We use fastqc to check read quality, and sickle for
#    trimming.

#    This rule requires to specify the following parameters: 
#         -MAX_MISMATCHES
#         -READS_ISO
#         -BOWTIE_INDEX

#     These parameters have to be defined in the snakefile (workflow).

# """

# # rule map_all:
# #     """Map trimmed reads for each sample"""
# #     input:expand('{read_dir}/{gsm}/{gsm}_bowtie.sam', gsm=GSM_LIST, read_dir=CHIP_READ_DIR)

# rule bowtie:
#     """Read mapping with bowtie version 1 (no gap, no paired-end)."""
#     input: "{trimmed_reads}.fastq"
#     output: "{trimmed_reads}_bowtie.sam"
#     log: "{trimmed_reads}_bowtie.log"
#     benchmark: "{trimmed_reads}_bowtie_benchmark.json"
#     params: max_mismatches = MAX_MISMATCHES, \
#             reads_iso = READS_ISO, \
#             index = BOWTIE_INDEX, \
#             qsub=config["qsub"]+" -e {trimmed_reads}_bowtie_qsub.err -o {trimmed_reads}_bowtie_qsub.out"
#     shell:"bowtie -q -S -n {params.max_mismatches} -m {params.reads_iso} --phred33-quals {params.index} {input} {output} 2> {log}"
