"""Annotate a genome (infer transcript regions) based on RNA-seq data.
Assemble transcripts from RNA-seq data and produce a file with
    the location of detected transcripts.

Author: Jacques.van-Helden@univ-amu.fr

Contributor : Claire Rioualen

TODO: doc + verifs

Params:
    threads
    libtype
    cuffmerge_dir
    assembly_files
    cuffmerge_transcripts

    genome/gtf_url


"""

# Genome annotations
genome_version = config["genome"]["version"]
genome_dir = config["dir"]["genome"] + genome_version + "/"
gtf_file = genome_dir + genome_version + ".gtf"



rule cufflinks:
    input:
        bam = "{reads}_sorted_pos.bam", \
        gtf = gtf_file
    output:
        outdir="{reads}_cufflinks", \
        transcripts = "{reads}_cufflinks/transcripts.gtf"
    log:  "{reads}_cufflinks.log"
    benchmark:  "{reads}_cufflinks_benchmark.json"
    params: libtype = config["cufflinks"]["libtype"], \
        threads = config["cufflinks"]["threads"], \
        qsub = config["qsub"] + " -e {reads}_cufflinks_qsub.err -o {reads}_cufflinks_qsub.out"
    shell: "cufflinks --GTF-guide {input.gtf} --num-threads {params.threads}  --library-type {params.libtype}  --output-dir {output.outdir} {input.bam} 2> {log}" 


