rule edgeR:
    """Run differential expression analysis on a table of counts per tag
    (gene). 

    This rule takes as input a tab-delimited text file containing gene
    counts per sample (one row per gene, one column per sample) and a gene
    annotation file (gtf).

    Result files:

    1. A tab-delimited file with with differential expression
       statistics, sorted by significance (p-value of FDR).
    2. A pdf file with various plots. 

    """
    input:  r_script = config["edgeR"]["script_R_path"], \
            R_params = PARAMS_R
    output: tab = RESULTS_EDGER
    log: config["dir"]["results"] + "DEG/edgeR.log"
    benchmark: config["dir"]["results"] + "DEG/edgeR_benchmark.json"
    params: qsub = config["qsub"] + " -e " + config["dir"]["results"] + "DEG/edgeR_qsub.err -o " + config["dir"]["results"] + "DEG/edgeR_qsub.out" #, \
#            cond1 = config["Diff_Exp"]["cond1"], \
#            cond2 = config["Diff_Exp"]["cond2"], \
#            dir_results = config["dir"]["results"]
    shell: "Rscript {input.r_script} {input.R_params}"
