rule HTseq_allcount_params:
    """
        Creates a file containing all count from given HTseq count file, 
        and extracts all useful data into a R_params file for differential expression 
        with edgeR and DESeq2
    """

    input:  COUNT_FILES
    output: R_params = PARAMS_R, \
            all_counts_file = ALL_COUNTS
    # log: "results/{cond_1}_VS_{cond_2}_bowtie2_mm" + config["bowtie2"]["max_mismatches"] + "_sorted_" + config["htseq"]["order"] + ".log"
    # benchmark: "results/{cond_1}_VS_{cond_2}_bowtie2_mm" + config["bowtie2"]["max_mismatches"] + "_sorted_" + config["htseq"]["order"] + ".json"
    params: qsub = config["qsub"] + " -q short", \
            cond1 = config["Diff_Exp"]["cond1"], \
            cond2=config["Diff_Exp"]["cond2"], \
            dir_results = config["dir"]["results"]

    run:
        # Create lists containing tested condition at the same index
        condition_1 = params.cond1
        condition_2 = params.cond2


        dir_results  = params.dir_results
    
        # Get name of R_params file without path
        params_R_file = output.R_params.split("/")[-1]

        # Get extention of align files from name of R_params
        ext_align = params_R_file.replace("_params.R", "")

        # Get extention of count file
        complet_path_count = COUNT_FILES[0].rsplit("/", maxsplit=1)
        count_ext = complet_path_count[-1].rsplit(config["htseq"]["order"])[-1]

        # Initilizing headers for all_count file
        header = "gene_id"
        # Add to header the name of the replicates in the same order as COUNT_FILES
        for count_file in COUNT_FILES:
            # Remove path of count table files
            head = count_file.split("/")[-1]
            header += "\t" + head.split("_" + ext_align)[0]

        # Condition only true on the first loop
        gene_ids_ok = True
        list_line = []

        #Parse a count files of all replicates of all conditions
        for i in range(len(COUNT_FILES)):
            count_res = open(COUNT_FILES[i])
            #Check if first loop
            if gene_ids_ok:
                for line in count_res:
                    # Create gene_ids
                    gene_counts_ids = line.split("\t")
                    list_line.append(gene_counts_ids[0] + "\t" + gene_counts_ids[1].strip("\n"))
                gene_ids_ok = False

            #If gene_ids already exist add count column
            list_file = count_res.readlines()
            for j in range(len(list_file)):
                gene_counts = list_file[j].split("\t")
                list_line[j] += "\t" + gene_counts[1].strip("\n")
            count_res.close()

        #Open and write all_count in a file
        all_counts = open(output.all_counts_file, 'w')
        all_counts.write(header + "\n")
        for elm in list_line:
            all_counts.write(elm + "\n")
        all_counts.close()


        # Initializing variables
        counts_f = []
        sample_condition_py = []
        sample_names_py = []
        conditions = []
        n_rep = []
        output_list = []

        #Get all parameters for all comparisons
        for i in range(len(condition_1)):

            cond_1 = condition_1[i]
            cond_2 = condition_2[i]
            output_list.append(dir_results + "DEG/" + cond_1 + "_vs_" + cond_2 + "/" + cond_1 + "_VS_" + cond_2 + ext_align + ".tab")
            

            #Get biological replicate list
            R_1 = config["Diff_Exp"][cond_1]
            R_2 = config["Diff_Exp"][cond_2]

            files_list = []
            files_list_R1 = []
            files_list_R2 = []
            # List path + name of replicats
            for i in range(len(R_1)) :
                files_list_R1.append('"' + config["dir"]["reads_source"] + R_1[i]+ "/" + R_1[i] + "_" + ext_align + count_ext + '"')
                files_list_R2.append('"' + config["dir"]["reads_source"] + R_2[i]+ "/" + R_2[i] + "_" + ext_align + count_ext + '"')

            files_list = files_list_R1 + files_list_R2
            
            #Separate the count files in string
            counts_f.append("c(" + ', \n '.join(files_list) + ")")
            #Get the list of condition 
            sample_condition_temp = ("'" + cond_1 + "',") * len(R_1) + ("'" +cond_2 + "',") * len(R_2)
            #remove the last coma
            sample_condition_py.append("c(" +  sample_condition_temp[:-1] + ")")

            #get list of replicate
            sample_names_py.append("c('" + "','".join(R_1) + "','" + "','".join(R_2) + "')")

            #list of condition
            conditions.append("c('" + cond_1 + "','" + cond_2 + "')")

            if len(R_1) >= len(R_2):
                n_rep.append(str(len(R_2)))
            else:
                n_rep.append(str(len(R_1)))
        
        out_R=open(output.R_params, 'w')
        out_R.write("## Root path \n" \
        + "data.root <-'" +  config["dir"]["results"] + "' \n \n" \
        + "## Table containing the counts of reads per gene (rows) for each sample (columns)  \n" \
        + "all.counts.table <- '" + output.all_counts_file + "' \n \n" \
        + "## Description of the conditions \n" \
        + "conditions <- c('" + "','".join(config["Diff_Exp"]["conditions"]) + "') \n" \
        + "n.rep <-c(" + ','.join(n_rep) + ")\n \n" \
        + "count.files <-c('" + "', \n '".join(COUNT_FILES) + "') \n \n"\
        + "FDR.threshold <- " + config["edgeR"]["FDR_threshold"] + "\n \n \n"\
        + "################################################################ \n" \
        + "## Structure of the comparisons for edgeR \n \n" \
        + "comparisons <-list(" + ','.join(conditions) + ")\n" \
        + "comparisons.cond1 <-list('" + "','".join(condition_1) + "') \n" \
        + "comparsions.cond2 <-list('" + "','".join(condition_2) + "') \n" \
        + "counts.files.per.comparisons <- list(" + ', \n'.join(counts_f) + ")\n" \
        + "condition.per.comparisons <-list(" + ', \n '.join(sample_condition_py) + ")\n" \
        + "names.per.comparisons <-list(" +  ', \n '.join(sample_names_py) + ")\n \n" \
        + "## List of cvs files that will be created by edgeR \n" \
        + "output <-c('" + "', \n '".join(output_list) + "') \n \n" \
        + "#### END OF THE CONFIG FILE \n" \
        + "################################################################" \
        )     
        out_R.close()