# Check if each required parameter has been defined in the JSON
# configuration file, otherwise use default values.

if not "size" in config["genome"].keys():
    sys.exit("The parameter genome size must be defined in the JSON file")

if not "macs2" in config.keys():
    config["macs2"] = {}

if not "qval" in config["macs2"].keys():
    config["macs2"]["qval"] = 0.05

if not "keep_dup" in config["macs2"].keys():
    config["macs2"]["keep_dup"] = "all"

if not "band_width" in config["macs2"].keys():
    config["macs2"]["band_width"] = 250

if not "mfold_min" in config["macs2"].keys():
    config["macs2"]["mfold_min"] = 2

if not "mfold_max" in config["macs2"].keys():
    config["macs2"]["mfold_max"] = 50

if not "call_summits" in config["macs2"].keys():
    config["macs2"]["call_summits"] = "--call-summits"

if not "other_options" in config["macs2"].keys():
    config["macs2"]["other_options"] = ""


rule macs2:
	"""Peak-calling with MACS2.
	Input: sam, bam, bed... automatically detected by macs' callpeak algorithm. Bed format was used here. 
	Output: bed

	/!\ Requires RSAT to be installed (maybe we should do this some other way to get rid of the dependency?)

	Required parameters:
		config['qsub']
		config["genome"]["size"]

	Usage: 
		PEAKS_MACS2 = expand(expand(RESULTS_DIR + "{treat}_vs_{ctrl}/macs2/{treat}_vs_{ctrl}{{trimming}}_{{aligner}}_macs2_peaks.narrowPeak",
	               zip, treat=TREATMENT, ctrl=CONTROL), trimming=config["sickle"]["suffix"], aligner=config["bwa"]["suffix"])
	"""
        input: 	treatment="{result_dir}/{treatment}/{treatment}" + config["suffixes"]["mapped"] + ".bed", \
            	control="{result_dir}/{control}/{control}" + config["suffixes"]["mapped"] + ".bed"
	output: narrowpeaks = "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_peaks.narrowPeak", \
                peaks_bed = "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_peaks.bed", \
                peak_len = "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_peaklen.tab", \
                summits = "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_summits.bed"
        log: "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2.log"
        benchmark: "{result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_benchmark.json"
        params: outdir="{result_dir}/{treatment}_vs_{control}/macs2/", \
            name="{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2", \
            qval = config["macs2"]["qval"], \
            call_summits = config["macs2"]["call_summits"], \
            keep_dup = config["macs2"]["keep_dup"], \
            band_width = config["macs2"]["band_width"], \
            mfold_min = config["macs2"]["mfold_min"], \
            mfold_max = config["macs2"]["mfold_max"], \
            other_options = config["macs2"]["other_options"], \
            genome_size = config["genome"]["size"], \
            qsub=config["qsub"] \
                  + " -q long -e {result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_qsub.err" \
                  + " -q long -e {result_dir}/{treatment}_vs_{control}/macs2/{treatment}_vs_{control}" + config["suffixes"]["mapped"] + "_macs2_qsub.out" 
	shell: "(macs2 \
callpeak -t {input.treatment} -c {input.control} --gsize {params.genome_size} \
--qvalue {params.qval} \
--keep-dup {params.keep_dup} \
--bw {params.band_width} \
--mfold {params.mfold_min} {params.mfold_max} \
--outdir {params.outdir} --name {params.name} \
{params.call_summits} {params.other_options} ; \
convert-features -from bed3col -to bed3col -i {output.narrowpeaks} -o {output.peaks_bed}; \
sequence-lengths -i {output.peaks_bed} -in_format bed | classfreq -ci 10 -v 1 -o {output.peak_len}) &> {log}"

