"""
    Trim reads required before alignment 

"""

rule trimming_paired_end:
    """Trim all raw paired-end read files (fastq.gz format) using sickle.

    The forward and the reverse read files have to be named
    respectively '{data_dir}_1.fastq.gz' and '{data_dir}_2.fastq.gz'.

    The quality threshold (threshold) and the quality type (seq_type)
    parameters have to be determined in the config file (json format)
    linked in the snakefile (workflow).

    You can add any other desired parameters in the "other options" of
    the config file.

    The result files are named '{data_dir}_1_trimmed.fastq.gz'
    (forward reads), '{reads}_2_trimmed.fastq.gz' (reverse reads) and
    '{data_dir}_trimmed.fastq.gz' (unmatched reads).

    """
    
    input: forward="{reads}" + SUFFIX_FWD + ".fastq.gz", reverse="{reads}" + SUFFIX_REV + ".fastq.gz"
    output: forward = "{reads}" + SUFFIX_FWD + "_trimmed_thr" + THRESHOLD + ".fastq.gz", \
            reverse = "{reads}" + SUFFIX_REV + "_trimmed_thr" + THRESHOLD + ".fastq.gz", \
            single = "{reads}_single_trimmed_thr" + THRESHOLD + ".fastq.gz", \
            summary = "{reads}_trimmed_thr" + THRESHOLD + "_summary.txt"
    params: threshold = THRESHOLD, seq_type=SEQ_TYPE, qsub=QSUB_PARAM + " -q long -e {reads}_trimmed_qsub.err -o {reads}_trimmed_qsub.out", options = OPTIONS_SICKLE
    log: "{reads}_trimmed_thr" + THRESHOLD + ".log"
    benchmark: "{reads}_trimmed_thr" + THRESHOLD + "_benchmark.json"
    # message: "Trimming raw reads with sickle\n\tInput:\t{input}\n\tOutput:\t{output}\n\tLog:\t{log}"
    shell: "sickle pe --gzip-output --pe-file1 {input.forward} --pe-file2 {input.reverse} --qual-type {params.seq_type} --qual-threshold {params.threshold} --output-pe1 {output.forward} --output-pe2 {output.reverse} --output-single {output.single} {params.options} > {output.summary} 2> {log}"
