"""
    Snakemake rule to order bam file by name of reads or position
    on the chromosome(s)
"""


rule sorted_bam:
    """
        Uses samtools to sort bam file and add "_sorted_[type_of_sort]"
        If you want to sort by position sort_by_name must be "" and config["htseq"]["order"] "pos".
        If you want to sort by name sort_by_name must be " -n" and config["htseq"]["order"] "name".

    """
    input: "{reads}.bam"
    output: "{reads}_sorted_" + config["htseq"]["order"] + ".bam"
    log: "{reads}_sorted_" + config["htseq"]["order"] + ".log"
    benchmark: "{reads}_sorted_" + config["htseq"]["order"] + "_benchmark.json"
    params: sort_by_name = config["samtools"]["sort_by_name"], \
    		qsub = config["qsub"]+" -e {reads}_sorted_qsub.err -o {reads}_sorted_qsub.out"
    shell:"samtools sort{params.sort_by_name} -f {input} {output} 2> {log}"


